sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/BatchRecorderSdk.java

```diff
@@ -10,61 +10,105 @@
 import io.opentelemetry.api.metrics.DoubleCounter;
 import io.opentelemetry.api.metrics.DoubleUpDownCounter;
 import io.opentelemetry.api.metrics.DoubleValueRecorder;
+import io.opentelemetry.api.metrics.Instrument;
 import io.opentelemetry.api.metrics.LongCounter;
 import io.opentelemetry.api.metrics.LongUpDownCounter;
 import io.opentelemetry.api.metrics.LongValueRecorder;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 /**
  * Minimal implementation of the {@link BatchRecorder} that simply redirects the calls to the
  * instruments.
- *
- * <p>TODO: Add an async queue processing to process batch records.
  */
 final class BatchRecorderSdk implements BatchRecorder {
   private final Labels labelSet;
 
+  private volatile Map<Instrument, List<Number>> recordings = new HashMap<>();
+  private volatile Map<Instrument, List<Number>> previousRecordings = new HashMap<>();
+
   BatchRecorderSdk(String... keyValuePairs) {
     this.labelSet = Labels.of(keyValuePairs);
   }
 
   @Override
   public BatchRecorder put(LongValueRecorder valueRecorder, long value) {
-    valueRecorder.record(value, labelSet);
+    recordings.computeIfAbsent(valueRecorder, instrument -> new ArrayList<>()).add(value);
```

Comment:
I'm guessing this can be called concurrently - so the maps need to be `ConcurrentHashMap` and the value something synchronized, probably `BlockingQueue`.

Reply:
Though given the primitive types have to be boxed anyways, I think wrapping in our own type would have very low overhead. So instead of two maps, I guess we can use a single queue? Would be easier to reason about.

```java
private static class LongRecording implements Recording {
  private final Instrument instrument;
  private final long value;
}

BlockingQueue<Recoding> recordings;

record() {
  recordings.drainTo(buffer);
  for (Recording recording : buffer) {
    ...
  }
```

Reply:
yeah, I went back and forth about this. I'd be happy to change it to something like that. And, yes, you're totally right about the concurrent hashmap bit.

Reply:
@anuraaga take a look now.